Initialize SciJava Concurrent module

gselzer opened the pull request:
This PR migrates @maarzt's fantastic parallelization work into a library independent of imglib2 - allowing us to use his work in a broader context throughout other SciJava libraries. This work is a culmination of the discussion in https://github.com/scijava/scijava/issues/31

TODO:
* [x] Is there a better name? As @ctrueden noted in https://github.com/scijava/scijava/issues/31, `scijava-parallel` is taken. We might want to do some renaming if we keep the name `scijava-concurrent`
* [x] @maarzt - can you add to the POM the `developer` roles you feel comfortable with? If you want add yourself as `author` in all of the classes, that'd be great too. I think the only one I changed significantly is `ParallelizationBenchmark.java`...
* [x] Similarly, @ctrueden you have full roles since I copied the POM from SciJava Types 😝. Do you want these?
* [x] We should decide if there are any classes in this module that should not be `export`ed. I think the package is small enough where it is probably unnecessary, but we might want to hide `ForkJoinExecutorService`/`SequentialExecutorService` away in some `org.scijava.concurrent.impl` package.

Closes https://github.com/scijava/scijava/issues/31

gselzer committed:
Scijava-concurrent: initial commit

Note that scijava-concurrent is a fork of the imglib2-parallel library -
see https://github.com/imglib/imglib2

gselzer committed:
Format classes

All of these classes used imglib2 styling, we needed to switch to
Scijava/ImageJ styling

gselzer commented:
I am unable to run `ParallelizationBenchmark.main` within Eclipse, getting 

```
WARNING: Unknown module: org.scijava.concurrent specified to --patch-module
WARNING: Unknown module: org.scijava.concurrent specified to --add-reads
Error: Could not find or load main class org.scijava.concurrent.ParallelizationBenchmark
Caused by: java.lang.ClassNotFoundException: org.scijava.concurrent.ParallelizationBenchmark
```

Can anyone reproduce this?

gselzer committed:
Use Parallelization in DefaultPValue

hinerm commented on imagej/imagej-ops2/src/main/java/net/imagej/ops2/coloc/pValue/DefaultPValue.java:
```diff
@@ -89,46 +89,38 @@ public void compute(final RandomAccessibleInterval<T> image1, final RandomAccess
+		};
+
+		try {
+			Parallelization.getTaskExecutor().forEach(params, task);
```
@gselzer Didn't you get the `executor` above on line 92?

gselzer replied:
Whoops, good catch! Fixed in f992fe5

gselzer committed:
Convert ImageJ Ops2 to use SciJava Parallel

gselzer committed:
Remove scijava-threads module

SciJava Concurrent is simply more powerful, and SciJava Threads was not
being used anywhere

gselzer commented:
> We should decide if there are any classes in this module that should not be exported. I think the package is small enough where it is probably unnecessary, but we might want to hide ForkJoinExecutorService/SequentialExecutorService away in some org.scijava.concurrent.impl package.

I think the answer to this is probably no - it seems to me like the user would want to choose their `ExecutorService` implementation, and they couldn't do that if we hid the implementations away. It's probably best to leave them exposed.

hinerm commented:
>We might want to do some renaming if we keep the name scijava-concurrent

@gselzer concurrent seems fine to me. The whole `scijava-parallel` thing is unfortunate. What would you rename though if we stick with concurrent?

gselzer commented:
> > We might want to do some renaming if we keep the name scijava-concurrent
> 
> @gselzer concurrent seems fine to me. The whole `scijava-parallel` thing is unfortunate. What would you rename though if we stick with concurrent?

I have no idea, it was too long ago - maybe I meant classes? Maybe we should rename `Parallelization` to `Concurrency`?

hinerm commented on imagej/imagej-ops2/src/main/java/net/imagej/ops2/filter/convolve/ConvolveNaiveF.java:
```diff
@@ -102,9 +102,9 @@ public RandomAccessibleInterval<O> apply(final RandomAccessibleInterval<I> input
-		// conforms only if the kernel is sufficiently small
-		if (Intervals.numElements(kernel) <= 9)
-			throw new IllegalArgumentException("The kernel is too small to perform computation!");
+//		// conforms only if the kernel is sufficiently small
```
@gselzer why is this commented out and not just deleted?

hinerm replied:
@gselzer why is this commented out and not just deleted?

hinerm commented on scijava/scijava-concurrent/pom.xml:
```diff
@@ -53,45 +53,63 @@
 	<ciManagement>
-		<system>GitHub Actions</system>
-		<url>https://github.com/scijava/incubator/actions</url>
+		<system>Travis CI</system>
```
@gselzer nope lol

hinerm commented on scijava/scijava-threads/README.md:
@gselzer should probably have a basic readme in the new concurrent module?

hinerm reviewed (changes requested):
I love this change.. getting rid of the `ExecutorService` is fantastic and the consuming code looks simpler in many cases.

Just a couple minor points. Thanks @gselzer!

gselzer committed:
Write a very basic README

ctrueden commented:
Not a full review from me, but I just wanted to chime in that I'm also pleased to see the explicit `ExecutorService` parameters go away in favor of the new `Parallelization` paradigm that tucks it neatly away while still enabling configurability. Thanks for working on it!

hinerm commented:
@gselzer looking good. Let's not rename things since this is a port of ImgLib2 code, basically.

Per our talk this morning we'll wait to merge until you add a tutorial in `imagej2-ops2-tutorial` showing its use

hinerm commented:
> We should decide if there are any classes in this module that should not be exported.

I wouldn't call these internal API so I am fine exporting everything in this module

gselzer committed:
Create Neighborhood parallel Ops

gselzer committed:
Write Parllel Ops tutorial

gselzer committed:
Write tutorial for non-Op parallel computation

gselzer commented:
@hinerm are all of your requested changes complete now?
